File: server/controllers/apis/barCodes.ts
import { verifyAdminSession } from './middlewares/sessions'

@Namespace('api', 'bar-codes')
@Middlewares({ before: [verifyAdminSession] })
export class BarCodesController {
  @Model('BarCodesModel') private barCodesModel: Models<'BarCodesModel'>
  @Get('/')
  public async index(_: PXIOHTTP.Request, res: PXIOHTTP.Response): Promise<void> {
    const results = await this.barCodesModel.getAll()
    res.json(results)
  }
  @Post('/')
  public async create(req: PXIOHTTP.Request, res: PXIOHTTP.Response): Promise<void> {
    const { name, tag = '', value } = req.body
    if (!name || !value) {
      res.json({
        ok: false,
        code: 'bad-request',
        message: 'No hay datos.'
      })
      return
    }
    await this.barCodesModel.create({ name, tag, value })
    res.json({
      ok: true
    })
  }
  @Put('/')
  public async update(req: PXIOHTTP.Request, res: PXIOHTTP.Response): Promise<void> {
    const { id, name, tag = '', value } = req.body
    if (!id || !name || !value) {
      res.json({
        ok: false,
        code: 'bad-request',
        message: 'No hay datos.'
      })
      return
    }
    await this.barCodesModel.update(Number(id), { name, tag, value })
    res.json({
      ok: true
    })
  }
  @Delete('/:id')
  public async delete(req: PXIOHTTP.Request, res: PXIOHTTP.Response): Promise<void> {
    const { id = '' } = req.params
    await this.barCodesModel.delete(Number(id))
    res.json({
      ok: true
    })
  }
}

File: server/api/barcodes.go
package api

import (
	"Miscellaneous/core/modules"
	"Miscellaneous/models/structs"
	"Miscellaneous/server/errors"
	"Miscellaneous/server/middlewares"
	"net/http"

	"github.com/labstack/echo/v4"
)

type BarCodesAPI struct{}

func (bc *BarCodesAPI) CreateBarCode(c echo.Context) error {
	var data structs.NewBarCode
	if err := c.Bind(&data); err != nil {
		return errors.ProcessError(&structs.CoreError{
			IsInternal: false,
			Code:       "missing-credentials",
			Message:    "Credenciales requeridas.",
		}, c)
	}

	err := modules.BarCodes.Create(data)
	if err != nil {
		return errors.ProcessError(err, c)
	}

	return c.NoContent(http.StatusAccepted)
}

func (bc *BarCodesAPI) GetBarCodes(c echo.Context) error {
	results, err := modules.BarCodes.GetAll()
	if err != nil {
		return errors.ProcessError(err, c)
	}

	return c.JSON(http.StatusOK, results)
}

func (bc *BarCodesAPI) UpdateBarCode(c echo.Context) error {
	var data structs.BarCode
	if err := c.Bind(&data); err != nil {
		return errors.ProcessError(&structs.CoreError{
			IsInternal: false,
			Code:       "missing-data",
			Message:    "Datos requeridos.",
		}, c)
	}

	err := modules.BarCodes.Update(data)
	if err != nil {
		return errors.ProcessError(err, c)
	}

	return c.NoContent(http.StatusAccepted)
}

func (bc *BarCodesAPI) DeleteBarCode(c echo.Context) error {
	id := c.Param("id")
	err := modules.BarCodes.Delete(id)
	if err != nil {
		return errors.ProcessError(err, c)
	}

	return c.NoContent(http.StatusAccepted)
}

func (bc *BarCodesAPI) GetBarCodeSrc(c echo.Context) error {
	id := c.Param("id")
	src, err := modules.BarCodes.GetSrc(id)
	if err != nil {
		return errors.ProcessError(err, c)
	}

	return c.String(http.StatusOK, src)
}

func RegisterBarcodesAPI(e *echo.Echo) {
	bc := BarCodesAPI{}

	e.POST("/api/bar-codes", bc.CreateBarCode, middlewares.SM.VerifySession)
	e.GET("/api/bar-codes", bc.GetBarCodes, middlewares.SM.VerifySession)
	e.PUT("/api/bar-codes", bc.UpdateBarCode, middlewares.SM.VerifySession)
	e.DELETE("/api/bar-codes/:id", bc.DeleteBarCode, middlewares.SM.VerifySession)
	e.GET("/api/bar-codes/:id", bc.GetBarCodeSrc, middlewares.SM.VerifySession)
}
